Integrate Muon optimizer (2725)

Saurabh750 opened the pull request:
#### Context
What is the purpose of this PR? Is it to
- [ ] add a new feature
- [ ] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

Please link to any issues this PR addresses.
#2725 

#### Changelog
What are the changes made in this PR?
* Integrating Muon optimizer as a pytorch implementation in torchtune.
* Modify recipes accordingly.

#### Test plan
Please make sure to do each of the following if applicable to your PR. If you're unsure about any one of these just ask and we will happily help. We also have a [contributing page](https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md) for some guidance on contributing.

- [ ] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [ ] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [ ] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [ ] run unit tests via `pytest tests`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [ ] manually run any new or modified recipes with sufficient proof of correctness
- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

#### UX
If your function changed a public API, please add a dummy example of what the user experience will look like when calling it.
Here is a [docstring example](https://github.com/pytorch/torchtune/blob/6a7951f1cdd0b56a9746ef5935106989415f50e3/torchtune/modules/vision_transformer.py#L285)
and a [tutorial example](https://pytorch.org/torchtune/main/tutorials/qat_finetune.html#applying-qat-to-llama3-models)

- [ ] I did not change any public API
- [ ] I have added an example to docs or docstrings

Saurabh750 commented:
Based on above comments, 2 things came to my mind:
1. As @krammnic suggested, I can implement a fused optimizer - a wrapper around Muon where the 2nd optimizer for linear layers will be of the choice of the user. This will eliminate the muon checks and will be a cleaner way.
2. If we want to provide more flexibility while assigning optimizer, we can do it on parameter level. 
Eg: Inside config file: 
optimizer:
  muon: [param1, param2]
  AdamW: []
muon:
  _component_: torchtune.modules.SingleDeviceMuon
  momentum: 0.95
  lr: 0.02
  weight_decay: 0
AdamW:
  _component_: bitsandbytes.optim.PagedAdamW
  lr: 1e-5

Muon will be assigned to only param1, param2 while AdamW will be assigned for all remaining ones.
Please do let me know your views, I'll implement whatever is suitable!

joecummings commented on torchtune/modules/muon.py:
```diff
@@ -0,0 +1,253 @@
+    return buf1c / (buf2c.sqrt() + eps)
+
+
+class MuonWithAuxAdam(torch.optim.Optimizer):
```
I think we should optimize for this interface so that users don't need to specify multiple optimizers in their code and we can remove the messy branching logic. 

My rough proposal is:

```python
class MuonWithAuxAdam(torch.optim.Optimizer):
    def __init__(
        self,
        params,
        *,
        muon_selector=None,
        muon_lr: float = 0.02,
        muon_momentum: float = 0.95,
        adam_lr: float = 3e-4,
        adam_betas=(0.9, 0.95),
        adam_eps: float = 1e-10,
        weight_decay: float = 0.0,
    ):        
       if muon_selector is None:
            muon_selector = (
                lambda name, p: p.ndim >= 2
                and "tok_embeddings" not in name
                and "output"        not in name
            )

        muon_params = [p for n, p in named_params if muon_selector(n, p)]
        adam_params = [p for n, p in named_params if not muon_selector(n, p)]

        muon_params.sort(key=lambda p: p.size(), reverse=True)

        super().__init__(
            [
                dict(params=muon_params,
                     lr=muon_lr,
                     momentum=muon_momentum,
                     weight_decay=weight_decay,
                     use_muon=True),
                dict(params=adam_params,
                     lr=adam_lr,
                     betas=adam_betas,
                     eps=adam_eps,
                     weight_decay=weight_decay,
                     use_muon=False),
            ],
            defaults={}
        )
```

What do you think?

Also cc @ebsmothers  for thoughts.

joecummings replied:
cc @krammnic, I think this is the easiest way to go. WDYT?

joecummings reviewed (commented):
This is awesome, thanks for porting this over!

The TL;DR of my thoughts is that we should optimize for the AuxAdam varities of the Muon optimizer so that our code stays clean and organized. I made some suggestions so that the instantiation of these classes is easy from code or config.

joecummings commented on torchtune/modules/muon.py:
```diff
@@ -0,0 +1,253 @@
+######################################################
```
Ah I see Mark maybe suggested two files, but I'd prefer to limit everything to `optim.py` until we see the need for new files. Our library has also gotten a little bloated.

Saurabh750 commented:
@joecummings @krammnic : I have added Muon to optim.py
I tried finetuning on alpaca dataset using qwen2-0.5B and compared AdamW and Muon.
Trained for 5, 10, 20 epochs for batchsize 5 and 10. 
For all the experiments, AdamW performed better than Muon. I am attaching snippet of one of the experiments:
![adamwvsMuon](https://github.com/user-attachments/assets/41a3774e-7809-4b4b-a82c-6e883c2ec1a9)

There is a custom implementation of Adam in Muon class. I tried using the existing pytorch implementation with the view to use any pre-existing implementations of optimizer for linear layers. But this will not be possible due to load_state_dict() which supports only single optimizer to be stored. But, I believe load_state_dict() in OptimizerInBackward supports multiple optimizers. Please correct me if I am wrong.

I have tried reducing the muon checks, but still there is a muon check in the main file. Also, I have updated the get_lr() method for returning the lr of Muon only. Please suggest if this is correct.

Saurabh750 commented:
Hi @joecummings @krammnic , I have updated the Muon optimizer implementation.
In below image: 
Green -> Muon optimizer for first 20 epochs with 5e-4 lr
Blue -> Muon optimizer for epochs 20 to 40 with 5e-5 lr
![image](https://github.com/user-attachments/assets/5cb7e0f4-bde6-448e-8e50-e0d9ef640368)

For below image:
Blue -> Muon optimizer for epochs 20 to 40 with 5e-5 lr
Red -> AdamW optimizer for epochs 0 to 20 with 2e-5 lr

![image](https://github.com/user-attachments/assets/431c74ff-95fe-4869-a73c-f860b93143d5)

As suggested earlier, switching to another implementation and playing around with HPO helped.
In the lr_scheduler, I am only returning the Muon learning rate. Please let me know if anything else needs attention.
